'client-executor' crate for performing retries on grpc clients; port Plugin Registry to this pattern

wimax-grapl opened the pull request:
Resurrecting Colin's pr here.
https://github.com/grapl-security/grapl/pull/1849

I cannot speak to literally anything in `client-executor` myself, but I can speak about the rust-proto macro change - a lot of it is not really a good fit for interfaces and is a much better fit for a repeatable macro. 

I haven't tackled the streaming RPC yet, but that's an edge case!

colin-grapl committed:
client-executor alone

colin-grapl committed:
Add circuit breaker to plugin-registry

wimax-grapl committed:
Introduce execute_client_rpc! macro to wrap all the executor stuff

wimax-grapl committed:
Clean up the create_plugin RPC a bit

wimax-grapl committed:
remove perform_healthcheck for plugin registry tests

jgrillo-grapl commented on src/rust/rust-proto/src/graplinc/grapl/api/plugin_registry/v1beta1_client.rs:
```diff
@@ -41,6 +43,15 @@ pub enum PluginRegistryServiceClientError {
 
+// This likely should be removed in favor of making all SerDe conversions
+// TryFrom<Error = SerDeError> instead of From.
+// Currently it's kind of scattershot.
```
It's not scattershot. The things that are fallible are `TryFrom` and the things that are infallible are `From`. If this code needs to paper over that, then this is a fine approach. I don't think we should compromise the clarity of the `From`/`TryFrom` distinction in the actual SerDe types to accommodate things like this though.

jgrillo-grapl commented on src/rust/rust-proto/src/client_macros.rs:
```diff
@@ -0,0 +1,38 @@
+/// It makes some expectations of the structure of `self` - like that it
+/// has a `self.executor` and `self.proto_client`.
+#[macro_export]
+macro_rules! execute_client_rpc {
```
Can we call this `execute_rpc` instead? The name `execute_client_rpc` sounds like it's executing a remote procedure call _against the client_ which doesn't really make sense. Instead, we're executing a remote procedure call against the server.

jgrillo-grapl reviewed (approved):
👍 just minor nits

wimax-grapl committed:
codeowners

wimax-grapl committed:
NotFound for when a row is, well, not found

wimax-grapl committed:
Add conditions to execution spawn thing

formatting

Okay, integrated the condition stuff into the client macro stuff

finishing touches

wimax-grapl committed:
change a comment jesse commented on

wimax-grapl committed:
formatting
